File: server/src/routes/reports.ts
import { Router } from 'express';
import { query } from '../database/index.js';
import { authenticate, AuthRequest } from '../middleware/auth.js';
import { ValidationError } from '../middleware/errorHandler.js';

const router = Router();

// Generate candidate evidence report
router.post('/candidate/:candidateId', authenticate, async (req: AuthRequest, res, next) => {
  try {
    const { candidateId } = req.params;

    // Get candidate data
    const candidateResult = await query('SELECT * FROM candidates WHERE id = $1', [candidateId]);
    if (candidateResult.rows.length === 0) {
      throw new ValidationError('Candidate not found');
    }

    const candidate = candidateResult.rows[0];

    // Get claims
    const claimsResult = await query(
      'SELECT * FROM claims WHERE candidate_id = $1',
      [candidateId]
    );

    // Get skills
    const skillsResult = await query(
      'SELECT * FROM skills WHERE candidate_id = $1',
      [candidateId]
    );

    // Get evidence
    const evidenceResult = await query(
      'SELECT * FROM evidence WHERE candidate_id = $1',
      [candidateId]
    );

    // Calculate statistics
    const supportedClaims = claimsResult.rows.filter((c: any) => c.evidence_status === 'Supported').length;
    const partiallySupportedClaims = claimsResult.rows.filter((c: any) => c.evidence_status === 'Partially Supported').length;
    const unverifiedClaims = claimsResult.rows.filter((c: any) => c.evidence_status === 'Unverified').length;
    const conflictingClaims = claimsResult.rows.filter((c: any) => c.evidence_status === 'Conflicting').length;

    const provenSkills = skillsResult.rows.filter((s: any) => s.is_proven).length;
    const totalSkills = skillsResult.rows.length;

    const sections = [
      {
        title: 'Candidate Overview',
        content: `${candidate.name} - ${candidate.detected_role}`,
        items: [`Email: ${candidate.email || 'Not provided'}`, `Location: ${candidate.location || 'Not provided'}`]
      },
      {
        title: 'Claims Analysis',
        content: `Total claims: ${claimsResult.rows.length}`,
        items: [
          `Supported: ${supportedClaims}`,
          `Partially Supported: ${partiallySupportedClaims}`,
          `Unverified: ${unverifiedClaims}`,
          `Conflicting: ${conflictingClaims}`
        ]
      },
      {
        title: 'Skills Verification',
        content: `Total skills: ${totalSkills}`,
        items: [
          `Proven skills: ${provenSkills}`,
          `Skills requiring evidence: ${totalSkills - provenSkills}`
        ]
      },
      {
        title: 'Evidence Summary',
        content: `Total evidence items: ${evidenceResult.rows.length}`,
        items: evidenceResult.rows.map((e: any) => `${e.evidence_type}: ${e.filename}`)
      }
    ];

    const reportResult = await query(
      `INSERT INTO reports (type, candidate_id, candidate_name, summary, sections)
      VALUES ($1, $2, $3, $4, $5)
      RETURNING *`,
      [
        'Candidate Proof Report',
        candidateId,
        candidate.name,
        `Evidence report for ${candidate.name} with ${supportedClaims} supported claims and ${provenSkills} proven skills`,
        JSON.stringify(sections)
      ]
    );

    res.json({
      success: true,
      data: {
        report: reportResult.rows[0],
        statistics: {
          supportedClaims,
          partiallySupportedClaims,
          unverifiedClaims,
          conflictingClaims,
          provenSkills,
          totalSkills,
          totalEvidence: evidenceResult.rows.length
        }
      }
    });
  } catch (error) {
    next(error);
  }
});

// Get report by ID
router.get('/:id', authenticate, async (req: AuthRequest, res, next) => {
  try {
    const { id } = req.params;
    const result = await query('SELECT * FROM reports WHERE id = $1', [id]);

    if (result.rows.length === 0) {
      throw new ValidationError('Report not found');
    }

    res.json({
      success: true,
      data: result.rows[0]
    });
  } catch (error) {
    next(error);
  }
});

export { router as reportsRouter };

File: backend/app/routes/reports.py
from fastapi import APIRouter, Depends, HTTPException, status
from sqlalchemy.orm import Session
import json
from app.database import get_db
from app.models.report import Report
from app.models.user import User
from app.models.claim import Claim
from app.models.skill import Skill
from app.models.evidence import Evidence
from app.models.candidate import Candidate
from app.schemas import ApiResponse
from app.auth import get_current_active_user

router = APIRouter()


@router.post("/candidate/{candidate_id}", response_model=ApiResponse)
async def generate_candidate_report(
    candidate_id: str,
    current_user: User = Depends(get_current_active_user),
    db: Session = Depends(get_db)
):
    # Get candidate data
    candidate = db.query(Candidate).filter(Candidate.id == candidate_id).first()
    if not candidate:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Candidate not found"
        )
    
    # Get claims
    claims = db.query(Claim).filter(Claim.candidate_id == candidate_id).all()
    
    # Get skills
    skills = db.query(Skill).filter(Skill.candidate_id == candidate_id).all()
    
    # Get evidence
    evidence = db.query(Evidence).filter(Evidence.candidate_id == candidate_id).all()
    
    # Calculate statistics
    supported_claims = len([c for c in claims if c.evidence_status == 'Supported'])
    partially_supported_claims = len([c for c in claims if c.evidence_status == 'Partially Supported'])
    unverified_claims = len([c for c in claims if c.evidence_status == 'Unverified'])
    conflicting_claims = len([c for c in claims if c.evidence_status == 'Conflicting'])
    
    proven_skills = len([s for s in skills if s.is_proven])
    total_skills = len(skills)
    
    sections = [
        {
            "title": "Candidate Overview",
            "content": f"{candidate.name} - {candidate.detected_role}",
            "items": [
                f"Email: {candidate.email or 'Not provided'}",
                f"Location: {candidate.location or 'Not provided'}"
            ]
        },
        {
            "title": "Claims Analysis",
            "content": f"Total claims: {len(claims)}",
            "items": [
                f"Supported: {supported_claims}",
                f"Partially Supported: {partially_supported_claims}",
                f"Unverified: {unverified_claims}",
                f"Conflicting: {conflicting_claims}"
            ]
        },
        {
            "title": "Skills Verification",
            "content": f"Total skills: {total_skills}",
            "items": [
                f"Proven skills: {proven_skills}",
                f"Skills requiring evidence: {total_skills - proven_skills}"
            ]
        },
        {
            "title": "Evidence Summary",
            "content": f"Total evidence items: {len(evidence)}",
            "items": [f"{e.evidence_type}: {e.filename}" for e in evidence]
        }
    ]
    
    new_report = Report(
        type="Candidate Proof Report",
        candidate_id=candidate_id,
        candidate_name=candidate.name,
        summary=f"Evidence report for {candidate.name} with {supported_claims} supported claims and {proven_skills} proven skills",
        sections=json.dumps(sections)
    )
    
    db.add(new_report)
    db.commit()
    db.refresh(new_report)
    
    return ApiResponse(
        success=True,
        data={
            "report": {
                "id": str(new_report.id),
                "type": new_report.type,
                "candidate_id": str(new_report.candidate_id),
                "candidate_name": new_report.candidate_name,
                "generated_at": new_report.generated_at.isoformat(),
                "summary": new_report.summary,
                "sections": json.loads(new_report.sections)
            },
            "statistics": {
                "supportedClaims": supported_claims,
                "partiallySupportedClaims": partially_supported_claims,
                "unverifiedClaims": unverified_claims,
                "conflictingClaims": conflicting_claims,
                "provenSkills": proven_skills,
                "totalSkills": total_skills,
                "totalEvidence": len(evidence)
            }
        }
    )


@router.get("/{report_id}", response_model=ApiResponse)
async def get_report(
    report_id: str,
    current_user: User = Depends(get_current_active_user),
    db: Session = Depends(get_db)
):
    report = db.query(Report).filter(Report.id == report_id).first()
    if not report:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Report not found"
        )
    
    return ApiResponse(
        success=True,
        data={
            "id": str(report.id),
            "type": report.type,
            "candidate_id": str(report.candidate_id),
            "candidate_name": report.candidate_name,
            "generated_at": report.generated_at.isoformat(),
            "summary": report.summary,
            "sections": json.loads(report.sections)
        }
    )
